### PR TITLE
sonobus: init at 1.6.2

### PR DESCRIPTION
**Pull request by PowerUser64**
Co-authored-by: @tobiasBora <tobias.bora@gmail.com>

Please note that this is my first contribution here, so I might have done something wrong. If there's something I need to do differently, please don't hesitate to let me know!

###### Description of changes

This PR adds SonoBus, an easy-to-use application for low-latency peer-to-peer audio streaming. Homepage: https://sonobus.net/

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PowerUser64** [comment]
@nixos/darwin this app is cross-platform, and should build on macos. It'd be great to get it working on macos if some people want to try it.

**Comment by tobiasBora** [comment]
Sounds good to me. That said, I'd be curious to see if it works on Darwin…  @NixOS/darwin-maintainers could someone check that? (PS: is it ok to ping you this way?)

File: pkgs/applications/audio/sonobus/default.nix
Comment (sikmir) [inline pkgs/applications/audio/sonobus/default.nix:79]:
```suggestion
    maintainers = with maintainers; [ PowerUser64 ];
    platforms = platforms.unix;
```

File: pkgs/applications/audio/sonobus/default.nix
Comment (sikmir) [inline pkgs/applications/audio/sonobus/default.nix:80]:
```suggestion
```
`mainProgram` is `pname` by default.

File: pkgs/applications/audio/sonobus/default.nix
Comment (sikmir) [inline pkgs/applications/audio/sonobus/default.nix:30]:
```suggestion
    repo = "sonobus";
    rev = version;
```

**Comment by sikmir** [comment]
> Sounds good to me. That said, I'd be curious to see if it works on Darwin… @NixOS/darwin-maintainers could someone check that? (PS: is it ok to ping you this way?)

`error: Package ‘alsa-lib-1.2.7.2’ in /nix/store/cad8sc7a3kiaq96mz5v35rmj5a405wqr-source/pkgs/os-specific/linux/alsa-project/alsa-lib/default.nix:36 is not supported on ‘x86_64-darwin’, refusing to evaluate.`

File: pkgs/applications/audio/sonobus/default.nix
Comment (totoroot) [inline pkgs/applications/audio/sonobus/default.nix:81]: Consider adding the meta attribute `platforms`.
If you want to package it for darwin as well, use:

```suggestion
  platforms  = platforms.unix;
  };
```

**Review by totoroot — CHANGES_REQUESTED** [review]

**Comment by tobiasBora** [comment]
@sikmir thanks! Btw any idea why ofborg-eval-darwin still passes? Is there anything we can do to ask ofborg to build on darwin? Or is it the test `sonobus, sonobus.passthru.tests on aarch64-darwin` that just appeared and is still running?

File: pkgs/applications/audio/sonobus/default.nix
Comment (PowerUser64) [reply pkgs/applications/audio/sonobus/default.nix:79]: I'm going to make this `platforms.unix` instead of just linux because it should work on darwin as well. (And I'm going to try and make it darwin-compatible.)

File: pkgs/applications/audio/sonobus/default.nix
Comment (wegank) [inline pkgs/applications/audio/sonobus/default.nix:80]:
```suggestion
    platforms = platforms.unix;
    broken = stdenv.isDarwin;
```

Comment (PowerUser64) [reply pkgs/applications/audio/sonobus/default.nix]: Good idea, thanks for the tip!

File: pkgs/applications/audio/sonobus/default.nix
Comment (lilyinstarlight) [inline pkgs/applications/audio/sonobus/default.nix:59]: I'm not sure if turning off patchelf has any significant side-effects. I generally see packages solve this by add a `wrapProgram` (from `makeWrapper` setup hook) in `installPhase` or `postInstall` to prefix `LD_LIBRARY_PATH`, but I've no strong opinions since I do see this construction elsewhere

The `libGL` and `libX*` deps could be moved to such a call (since they aren't needed at build-time) if you ended up preferring that:

```
    wrapProgram $out/bin/sonobus \
      --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ libGL libX11 libXcursor libXext libXinerama libXrandr ]}"
```

File: pkgs/applications/audio/sonobus/default.nix
Comment (lilyinstarlight) [inline pkgs/applications/audio/sonobus/default.nix:72]:
```suggestion
    runHook preInstall
    cd ../linux
    ./install.sh "$out"
    runHook postInstall
```

Don't forget to run hooks if you set a custom phase

**Review by lilyinstarlight — COMMENTED** [review]
I left a few comments and questions, but this looks pretty good!

Most of the commits should be squashed into the first commit before this should be merged and commit 64365da1be476be9f420c77ae5af2e76fb827fb8 should still be separate but have description changed to say `maintainers: add PowerUser64` to fit contributing guidelines

**Comment by PowerUser64** [comment]
Since I wrote this, sonobus has made a new release, v1.6.2. @lilyinstarlight would it be okay for me to update the PR to 1.6.2?
(not sure if I'm allowed to do that)

**Comment by lilyinstarlight** [comment]
> Since I wrote this, sonobus has made a new release, v1.6.2. @lilyinstarlight would it be okay for me to update the PR to 1.6.2? (not sure if I'm allowed to do that)

Yes, please go ahead and update this PR! No reason to merge an already-outdated package :)

**Comment by PowerUser64** [comment]
Alright, I think this is ready to be merged.

**Review by lilyinstarlight — APPROVED** [review]
Aside from the concern about `dontPatchELF`, I don't see any issues with this PR so I would say it's mergable now

File: pkgs/applications/audio/sonobus/default.nix
Comment (tobiasBora) [reply pkgs/applications/audio/sonobus/default.nix:56]: Well as I understand rpath is supposed to be cleaner than `LD_LIBRARY_LOAD` no? (e.g. it works with setuid binaries, don't need wrapper…). It's just that by default nix over-optimizes the rpath for some applications.

Comment (lilyinstarlight) [reply pkgs/applications/audio/sonobus/default.nix:56]: Yeah but `dontPatchELF` prevents it from removing a lot of unnecessary runtime references. See the difference in closure size below:

```
[nix-shell:~/.cache/nixpkgs-review/pr-211341/nixpkgs]$ nix path-info -Sh /nix/store/p32vnnl6pw72zs53m9c3w0qpbbpk5q1q-sonobus-1.6.2  # straight from this PR as of d0ad888a4
/nix/store/p32vnnl6pw72zs53m9c3w0qpbbpk5q1q-sonobus-1.6.2	 593.3M

[nix-shell:~/.cache/nixpkgs-review/pr-211341/nixpkgs]$ nix path-info -Sh /nix/store/c1w40v5wfkva9l5b1vi2i9q3ybbxrqir-sonobus-1.6.2  # with my suggested change
/nix/store/c1w40v5wfkva9l5b1vi2i9q3ybbxrqir-sonobus-1.6.2	  97.2M
```

If we want to set rpath manually (e.g. if it does turn out to be needed for the vst plugins, which I didn't test), I would add it in a `postFixup` like `patchelf --add-rpath ${lib.makeLibraryPath [ libGL libX11 libXcursor libXext libXinerama libXrandr ]} $out/bin/sonobus $out/lib/vst3/*/Contents/x86_64-linux/*.so` and remove them from `buildInputs` since the cross splicing would still resolve the target deps correctly

Comment (lilyinstarlight) [reply pkgs/applications/audio/sonobus/default.nix:56]: Yeah I am fine if it just does have to be an rpath over all of the inputs (the amount I actually care is probably less than I made it sound), but if we want to do this optimization, it could be convenient for some

I'm preparing a diff in a minute to explain exactly how to patch in an additional rpath to stuff

Comment (tobiasBora) [reply pkgs/applications/audio/sonobus/default.nix:56]: Ok thanks. Just, my reason to prefer `NIX_LDFLAGS = "-rpath…"` over manually calling `patchelf` is that it is easy with the last method to forget some files to patch (especially if the library is getting new plugins format later).

Comment (lilyinstarlight) [reply pkgs/applications/audio/sonobus/default.nix:56]: This is the diff. I keep forgetting that `autoPatchelfHook` automates this so it's actually pretty simple

```diff
diff --git a/pkgs/applications/audio/sonobus/default.nix b/pkgs/applications/audio/sonobus/default.nix
index 26c6a0f0504..da82fcc5a86 100644
--- a/pkgs/applications/audio/sonobus/default.nix
+++ b/pkgs/applications/audio/sonobus/default.nix
@@ -2,6 +2,7 @@
 , pkg-config
 , stdenv
 , fetchFromGitHub
+, autoPatchelfHook
 , alsa-lib
 , cmake
 , freetype
@@ -31,6 +32,7 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [
+    autoPatchelfHook
     cmake
     pkg-config
   ];
@@ -38,12 +40,6 @@ stdenv.mkDerivation rec {
   buildInputs = [
     alsa-lib
     freetype
-    libGL
-    libX11
-    libXcursor
-    libXext
-    libXinerama
-    libXrandr
     libjack2
     libopus
     curl
@@ -51,9 +47,14 @@ stdenv.mkDerivation rec {
     webkitgtk
   ];
 
-  # To add the libraries to the rpath, needed or the program segfaults
-  NIX_LDFLAGS = "-rpath ${lib.makeLibraryPath buildInputs}";
-  dontPatchELF = true;
+  runtimeDependencies = [
+    libGL
+    libX11
+    libXcursor
+    libXext
+    libXinerama
+    libXrandr
+  ];
 
   postPatch = lib.optionalString (stdenv.isLinux) ''
     # needs special setup on Linux, dunno if it can work on Darwin
```

Comment (tobiasBora) [reply pkgs/applications/audio/sonobus/default.nix:56]: Oh thanks, I thought autopatchelf would only work for pre-built binaries, good to know! Maybe the wrapper adds bash to the closure (unless you used the binary version).

Comment (lilyinstarlight) [reply pkgs/applications/audio/sonobus/default.nix:56]: Oh yeah it probably was bash. I can agree with you that wrappers suck, though, so I am glad the autoPatchelf hook makes it very idiomatic to express and manages to catch all shared objects and binaries too

**Review by lilyinstarlight — APPROVED** [review]
That `dontPatchELF` concern has now been fixed (reducing closure size quite a lot), so this PR looks fantastic to me now!

File: pkgs/applications/audio/sonobus/default.nix
Comment (viraptor) [inline pkgs/applications/audio/sonobus/default.nix:41]: In the future, it should be possible to build a darwin version by making alsa optional... but that's a battle for another day.

**Review by viraptor — APPROVED** [review]

**Pull request merged**
